File: sdk/python/src/axcp/envelope.py
"""AXCP envelope serialization, signing, and verification."""

from __future__ import annotations

import base64
import uuid

from google.protobuf.message import DecodeError

from axcp.errors import InvalidEnvelopeError, SignatureMissingError
from axcp.identity import (
    DID,
    DIDResolver,
    Identity,
    public_key_from_did_key,
    resolve_ed25519_public_key,
    verify_signature,
)
from axcp.pb import axcp_pb2
from axcp.timestamp import now_ms, rfc3339_seconds_from_ms

DID_AUTH_TRANSCRIPT_VERSION = "AXCP-DID-AUTH-v1"
DEFAULT_PROTOCOL_VERSION = 1
DEFAULT_PROFILE_SECURE_BASELINE = 1


def new_envelope(
    trace_id: str | None = None,
    profile: int = DEFAULT_PROFILE_SECURE_BASELINE,
) -> axcp_pb2.AxcpEnvelope:
    env = axcp_pb2.AxcpEnvelope()
    env.version = DEFAULT_PROTOCOL_VERSION
    env.trace_id = trace_id or str(uuid.uuid4())
    env.profile = profile
    return env


def encode_envelope(envelope: axcp_pb2.AxcpEnvelope) -> bytes:
    return envelope.SerializeToString(deterministic=True)


def decode_envelope(data: bytes) -> axcp_pb2.AxcpEnvelope:
    env = axcp_pb2.AxcpEnvelope()
    try:
        env.ParseFromString(data)
    except DecodeError as exc:
        raise InvalidEnvelopeError("failed to decode AXCP envelope") from exc
    return env


def signing_payload(envelope: axcp_pb2.AxcpEnvelope) -> bytes:
    """Return deterministic protobuf payload covered by AXCP DID auth."""

    clone = axcp_pb2.AxcpEnvelope()
    clone.CopyFrom(envelope)
    clone.sender_did = ""
    clone.recipient_did = ""
    clone.timestamp_ms = 0
    clone.sequence = 0
    clone.ClearField("signature")
    clone.ClearField("attestation_proof")
    return clone.SerializeToString(deterministic=True)


def build_auth_transcript(envelope: axcp_pb2.AxcpEnvelope) -> bytes:
    payload = signing_payload(envelope)
    payload_b64 = base64.b64encode(payload).decode("ascii")
    timestamp = rfc3339_seconds_from_ms(envelope.timestamp_ms)
    return "\n".join(
        [
            DID_AUTH_TRANSCRIPT_VERSION,
            envelope.sender_did,
            envelope.recipient_did,
            payload_b64,
            timestamp,
        ]
    ).encode("utf-8")


def sign_envelope(
    envelope: axcp_pb2.AxcpEnvelope,
    identity: Identity,
    *,
    recipient_did: str = "",
    sequence: int | None = None,
    timestamp_ms: int | None = None,
) -> axcp_pb2.AxcpEnvelope:
    if timestamp_ms is not None and timestamp_ms < 0:
        raise InvalidEnvelopeError("timestamp_ms cannot be negative")
    if sequence is not None and sequence < 0:
        raise InvalidEnvelopeError("sequence cannot be negative")

    envelope.sender_did = identity.did
    envelope.recipient_did = recipient_did
    envelope.timestamp_ms = timestamp_ms if timestamp_ms is not None else now_ms()
    if sequence is not None:
        envelope.sequence = sequence
    transcript = build_auth_transcript(envelope)
    envelope.signature = identity.sign(transcript)
    return envelope


def verify_envelope(
    envelope: axcp_pb2.AxcpEnvelope,
    *,
    public_key: bytes | None = None,
    resolver: DIDResolver | None = None,
) -> None:
    if not envelope.signature:
        raise SignatureMissingError("envelope signature is missing")
    if not envelope.sender_did:
        raise InvalidEnvelopeError("sender_did is required")

    resolved_public_key = public_key or resolve_ed25519_public_key(envelope.sender_did, resolver)
    parsed_sender = DID.parse(envelope.sender_did)
    if (
        public_key is not None
        and parsed_sender.method == "key"
        and public_key_from_did_key(envelope.sender_did) != public_key
    ):
        raise InvalidEnvelopeError("explicit public key does not match sender did:key")
    transcript = build_auth_transcript(envelope)
    verify_signature(resolved_public_key, transcript, bytes(envelope.signature))

File: sdk/typescript/src/envelope.ts
import { randomUUID } from "node:crypto";
import Long from "long";
import { encodeBase64, toUint8Array, utf8Bytes } from "./bytes.js";
import {
  InvalidEnvelopeError,
  SignatureMissingError,
} from "./errors.js";
import {
  DID,
  type DIDResolver,
  type Identity,
  publicKeyFromDidKey,
  resolveEd25519PublicKey,
  verifySignature,
} from "./identity.js";
import { type AxcpEnvelope, AxcpEnvelopeType, toSafeNumber } from "./pb/schema.js";
import { nowMs, rfc3339SecondsFromMs } from "./timestamp.js";

export const DID_AUTH_TRANSCRIPT_VERSION = "AXCP-DID-AUTH-v1";
export const DEFAULT_PROTOCOL_VERSION = 1;
export const DEFAULT_PROFILE_SECURE_BASELINE = 1;

export interface NewEnvelopeOptions {
  readonly traceId?: string;
  readonly profile?: number;
}

export interface SignEnvelopeOptions {
  readonly recipientDid?: string;
  readonly sequence?: number;
  readonly timestampMs?: number;
}

export interface VerifyEnvelopeOptions {
  readonly publicKey?: Uint8Array;
  readonly resolver?: DIDResolver;
}

export function newEnvelope(options: NewEnvelopeOptions = {}): AxcpEnvelope {
  return {
    version: DEFAULT_PROTOCOL_VERSION,
    traceId: options.traceId ?? randomUUID(),
    profile: options.profile ?? DEFAULT_PROFILE_SECURE_BASELINE,
  };
}

export function encodeEnvelope(envelope: AxcpEnvelope): Uint8Array {
  const normalized = stripProto3Defaults(envelope) as AxcpEnvelope;
  const error = AxcpEnvelopeType.verify(normalized);
  if (error !== null) {
    throw new InvalidEnvelopeError(`invalid AXCP envelope: ${error}`);
  }
  return AxcpEnvelopeType.encode(AxcpEnvelopeType.create(normalized)).finish();
}

export function decodeEnvelope(data: Uint8Array): AxcpEnvelope {
  try {
    return AxcpEnvelopeType.decode(data) as unknown as AxcpEnvelope;
  } catch (error) {
    throw new InvalidEnvelopeError("failed to decode AXCP envelope");
  }
}

export function signingPayload(envelope: AxcpEnvelope): Uint8Array {
  const clone = decodeEnvelope(encodeEnvelope(envelope)) as Record<string, unknown>;
  delete clone.senderDid;
  delete clone.recipientDid;
  delete clone.timestampMs;
  delete clone.signature;
  delete clone.attestationProof;
  return encodeEnvelope(clone as AxcpEnvelope);
}

export function buildAuthTranscript(envelope: AxcpEnvelope): Uint8Array {
  const payload = signingPayload(envelope);
  const payloadB64 = encodeBase64(payload);
  const timestamp = rfc3339SecondsFromMs(toSafeNumber(envelope.timestampMs, "timestampMs"));
  return utf8Bytes(
    [
      DID_AUTH_TRANSCRIPT_VERSION,
      envelope.senderDid ?? "",
      envelope.recipientDid ?? "",
      payloadB64,
      timestamp,
    ].join("\n"),
  );
}

export function signEnvelope(
  envelope: AxcpEnvelope,
  identity: Identity,
  options: SignEnvelopeOptions = {},
): AxcpEnvelope {
  if (options.timestampMs !== undefined && options.timestampMs < 0) {
    throw new InvalidEnvelopeError("timestampMs cannot be negative");
  }
  if (options.sequence !== undefined && options.sequence < 0) {
    throw new InvalidEnvelopeError("sequence cannot be negative");
  }

  envelope.senderDid = identity.did;
  envelope.recipientDid = options.recipientDid ?? "";
  envelope.timestampMs = options.timestampMs ?? nowMs();
  if (options.sequence !== undefined) {
    envelope.sequence = options.sequence;
  }
  envelope.signature = identity.sign(buildAuthTranscript(envelope));
  return envelope;
}

export async function verifyEnvelope(envelope: AxcpEnvelope, options: VerifyEnvelopeOptions = {}): Promise<void> {
  const signature = toUint8Array(envelope.signature);
  if (signature.length === 0) {
    throw new SignatureMissingError("envelope signature is missing");
  }
  if (!envelope.senderDid) {
    throw new InvalidEnvelopeError("sender_did is required");
  }

  const resolvedPublicKey = options.publicKey ?? (await resolveEd25519PublicKey(envelope.senderDid, options.resolver));
  const parsedSender = DID.parse(envelope.senderDid);
  if (
    options.publicKey !== undefined &&
    parsedSender.method === "key" &&
    !bytesMatch(publicKeyFromDidKey(envelope.senderDid), options.publicKey)
  ) {
    throw new InvalidEnvelopeError("explicit public key does not match sender did:key");
  }

  verifySignature(resolvedPublicKey, buildAuthTranscript(envelope), signature);
}

function bytesMatch(left: Uint8Array, right: Uint8Array): boolean {
  if (left.length !== right.length) {
    return false;
  }
  let diff = 0;
  for (let index = 0; index < left.length; index += 1) {
    diff |= left[index]! ^ right[index]!;
  }
  return diff === 0;
}

function stripProto3Defaults(value: unknown): unknown {
  if (value === undefined || value === null) {
    return undefined;
  }
  if (typeof value === "string") {
    return value.length === 0 ? undefined : value;
  }
  if (typeof value === "number") {
    return value === 0 ? undefined : value;
  }
  if (typeof value === "bigint") {
    return value === 0n ? undefined : value;
  }
  if (typeof value === "boolean") {
    return value ? value : undefined;
  }
  if (Long.isLong(value)) {
    return value.isZero() ? undefined : value;
  }
  if (value instanceof Uint8Array) {
    return value.length === 0 ? undefined : value;
  }
  if (Array.isArray(value)) {
    const normalizedItems = value
      .map((item) => stripProto3Defaults(item))
      .filter((item): item is NonNullable<unknown> => item !== undefined);
    return normalizedItems.length === 0 ? undefined : normalizedItems;
  }
  if (typeof value === "object") {
    const normalized: Record<string, unknown> = {};
    for (const [key, item] of Object.entries(value)) {
      const normalizedItem = stripProto3Defaults(item);
      if (normalizedItem !== undefined) {
        normalized[key] = normalizedItem;
      }
    }
    return normalized;
  }
  return value;
}
